libminifi/include/utils/ChecksumCalculator.h

```diff
@@ -0,0 +1,57 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <string>
+#include <utility>
+
+#include "utils/OptionalUtils.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+
+class ChecksumCalculator {
+ public:
+  static constexpr const char* CHECKSUM_TYPE = "SHA256";
+  static constexpr size_t LENGTH_OF_HASH_IN_BYTES = 32;
+
+  void setFileLocation(const std::string& file_location);
```

Comment:
The class implementation could be simplified by requiring a valid file location at all times (setting the member in the constructor). If the usage side needs late initialization, it could rely on an optional or unique_ptr.

Reply:
`ChecksumCalculator` would be simpler, but its usage would be more complex.  In both Properties.h and FlowConfiguration.h, instead of `{ return checksum_calculator_; }`, we would need to check if the optional is valid, and construct the object if it isn't.

Reply:
We still return a useless object before it's initialized, it would just be a null pointer or other representation of "nothing" instead of a useless `ChecksumCalculator` that throws on calling `getChecksum`.

Reply:
I could change the `if (!...) { throw ...}` to `gsl_Expects`.  That would express the logic more clearly, but then these lines could no longer be unit tested, as the test would terminate instead of throwing.  Do you think that would be better?

Reply:
I'm fine with not testing invalid calls if the preconditions are documented. We don't need to test what happens on `gsl_Expects(false)`.

By "the getter could have a precondition that the appropriate member must be initialized before calling it", I meant the getter of `checksum_calculator_` in FlowConfiguration.h and Properties.h, but I'm not against a similar change in `ChecksumCalculator` either.

Reply:
Done (at least partially) in https://github.com/apache/nifi-minifi-cpp/pull/1123/commits/e83ec43a61ce4cf1e4bbac98451a8c8f20ba0982.